### ISSUE TITLE
Optimization to use ECDH-1PU only on first setup and use 'dir' afterwards

### ISSUE THREAD

**Issue by kdenhartog** [decentralized-identity/didcomm-messaging#255]
For along time now I've expected that in many cases that for a few of the steps in the ECDH-1PU algorithm that we could actually cache the results of the operations in order to optimize portions of ECDH-1PU. However, we could just as easily make the optimization go further within the spec to say the first key used between two keys (e.g. did:example:123#sender and did:example:123#recipient) that they take the cek from that handshake agreement and switch over to the 'dir' algorithm for all subsequent messages. This CEK could be used up to a certain arbitrary number or until a party chooses to reset the session by sending a new ECDH-1PU in which case the CEK gets rotated.

There's definitely some additional compexity that's introduced at the crypto layer to make this happen but it seems possible. What's peoples thoughts on this?

Thinking @ashcherbakov @AnomalRoil and @vimmerru will want to weigh in here

**Comment by ashcherbakov** [comment]
Technically it's possible, and it may be a must-have feature when DIDComm needs to be implemented on embedded devices where there can be not enough power to perform multiple key agreements for every message (FYI @dhh1128 ).

As for non-embedded case, I believe it was discussed somewhere that DIDComm is intentionally stateless, that's why it explicitly requires to perform a new key agreement and use a new CEK for every new message.
https://identity.foundation/didcomm-messaging/spec/#perfect-forward-secrecy is probably also important.

**Comment by dhh1128** [comment]
I agree that this is important. The complexity might not be a problem if we can easily code it into the libraries that DSR is building. However, we would have to be VERY careful to analyze the security properties of the crypto.

Re. embedded: I agree with Alex's comment. This worries me. We need embedded devices to be able to exchange messages in milliseconds, but I have heard estimates that it will take 3-5 seconds instead. That will never lead to adoption. So let's explore.

**Comment by kdenhartog** [comment]
Agreed that we'll need to assess the security properties. The concerns we'll need to make sure we give a look at is going to be the number of messages, how nonces are handled, and which cipher is being used (e.g. AES-GCM vs chacha20poly1305 have different needs to remain secure)

**Comment by kdenhartog** [comment]
@TelegramSam can we get this issue on the agent to decide if it's within scope of v2 or should be deferred?

**Comment by AnomalRoil** [comment]
Notice this is somewhat related to #243 as it would basically introduce the notion of session.

Also this would require being somewhat stateful since we'd need to keep track of the CEK currently used to communicate with XY.

From a security pov, it reduces forward secrecy guarantees to each such "session", which is usually fine, and it requires setting a max quantity of data we can encrypt with one CEK to avoid IV re-use issues. (Notice that the security currently doesn't depend on the IV, we could use a hash of the CEK, or a static IV since we use a different key each time. This wouldn't be true anymore)
